mozilla/bedrock
# Provide a hybrid list of locales drawn from CMS-backed and Django-backed sources

## One-line summary

This changeset ensures that the language picker (and `canonical-url.html` header partial) remains accurate when a page is partly in Wagtail CMS and partly rendered via Django views.

- [ ] I used an AI to write some of this code.

## Significant changes and points to review

This piece of work was like pulling at a loose thread on a sweater. It may look like there's a lot going on here, so I'll add notes on the Github diff where I think they'll help.

The `prefer_cms` decorator is applied to a regular Django view and will try to load the decorated URL path from the CMS first, then fall back to the original Django view. Previously, when a page came from the CMS, we would set the `translations` context variable to be a list of locales _available in the CMS only_, which had the downside that the language picker in the footer didn't mention the locales that were still available via the regular Django view. (The inverse was also true: if you reached a Django-rendered page that was decorated with `prefer_cms`, it would not show the locales available in the CMS because it didn't have awareness of them.

I tried different ways to get around this, including patching `translations` in `lib.l10n_utils.render` (which ended very badly as `translations` is used for more than just the footer picker, such as the logic around deciding whether to redirect to the default `en-US` locale.)

The approach that worked better was to, via the `prefer_cms` decorator, annotate the request with two lists of locale language codes: the locales available from the CMS and those available from the Django view. Then, we use a new Jinja helper to pull together these two lists for use in the language picker (we also use the same helper to provide correct locale lists for the canonical URL links in the header). 

(Note that this logic only runs on views decorated with `prefer_cms` - otherwise we still use the regular `translations` context var for such things.)

To make this work the `prefer_cms` decorator needs to be told what locales are available in the regular Django view - it was not viable to introspect the view to get hold of the relevant locales. There are three (!) ways to pass this information in:

1) Pass in the same Fluent files that are used in the view – this is good for views that are HTML templates and Fluent strings _only_. We work out the relevant locales based on the same get_active_locales()` logic as our main `l10n_utils.render()` function.
2) Pass in a list of language codes for the locales - this is good for views that might have content from an external source (eg a git repo) but which we know will be consistentely and completely translated into all of those locales
3) Pass in a callable function or method that will return the appropriate locales for the Django-rendered view. This is best for views with a single URL featuring a slug variable, where _some_ pages are widely localized, but not all of them are, so we can't reliably used a fixed list of locales. Things like the old VPN Resource Center articles are a good example of this, and **this changeset uses the callable pattern to make the language picker work for the VPN RC*

## Issue / Bugzilla link

Resolves #15193 

## Testing

* Use the VPN RC as the test for this. (See `Testing` section on https://github.com/mozilla/bedrock/pull/15236 if you need to set one up locally) 
* Add new CMS pages to the VPN RC using slugs that exist for the Django view in more than just `en-US` (e.g. `what-is-a-vpn` and `the-difference-between-a-vpn-and-a-web-proxy`) - they can just have lorem ipsum content locally - the URL loath is what we are really need here
* Publish the pages and view them.
* Edit bedrock/product.urls
* The footer's language picker should allow you to switch between any of the 10 languages available for those pages regardless of whether they are in the CMS or in Django, with no duplicated locales mentioned and never a 404 or redirect back to the `en-US` version
* The canonical url links in the `head` of the pages should also reflect the other locales available.
* Note that the the VPN RC index page in the CMS currently does _not_ mix Django and CMS-based content - that change is out of scope of this fix, which is focused only on the footer lang switcher
